engines/pink/cel_decoder.cpp

```diff
@@ -0,0 +1,225 @@
+	readHeader();
+}
+
+#define PREFIX_TYPE 0xF100
+#define CEL_DATA 3
+
+void CelDecoder::CelVideoTrack::readPrefixChunk() {
+	_fileStream->seek(0x80);
+	/* uint32 chunkSize = */_fileStream->readUint32LE();
+	uint16 chunkType = _fileStream->readUint16LE();
+	if (chunkType != PREFIX_TYPE)
+		return;
+	//uint32 offset = 6;
+
+	uint32 subchunkSize = _fileStream->readUint32LE();
+	uint16 subchunkType = _fileStream->readUint16LE();
+
+	switch (subchunkType) {
+	case CEL_DATA:
+		_fileStream->readUint16LE();
+		_center.x = _fileStream->readUint16LE();
+		_center.y = _fileStream->readUint16LE();
+		break;
+	default:
+		error("Unknown subchunk type");
+		_fileStream->skip(subchunkSize - 6);
+		break;
+	}
+}
+
+void CelDecoder::CelVideoTrack::readHeader() {
+	_fileStream->readUint16LE();
+
+	_frameDelay = _startFrameDelay = _fileStream->readUint32LE();
+
+	_fileStream->seek(80);
+	_offsetFrame1 = _fileStream->readUint32LE();
+	_offsetFrame2 = _fileStream->readUint32LE();
+
+	if (_offsetFrame1 > 0x80)
+		readPrefixChunk();
+
+	_fileStream->seek(_offsetFrame1);
+}
+
+uint16 CelDecoder::CelVideoTrack::getTransparentColourIndex() {
+	return _transparentColourIndex;
+}
+
+const Graphics::Surface *CelDecoder::CelVideoTrack::getCurrentFrame() {
+	return _surface;
+}
+
+Common::Point CelDecoder::CelVideoTrack::getCenter() {
+	return _center;
+}
+
+#define FRAME_TYPE 0xF1FA
+
+void CelDecoder::CelVideoTrack::skipFrame() {
+	// Read chunk
+	/*uint32 frameSize = */ _fileStream->readUint32LE();
+	uint16 frameType = _fileStream->readUint16LE();
+
+	switch (frameType) {
+	case FRAME_TYPE:
+		handleFrame();
+		break;
+	default:
+		error("FlicDecoder::decodeFrame(): unknown main chunk type (type = 0x%02X)", frameType);
+		break;
+	}
+
+	_curFrame++;
+	//_nextFrameStartTime += _frameDelay;
+
+	if (_atRingFrame) {
+		// If we decoded the ring frame, seek to the second frame
+		_atRingFrame = false;
+		_fileStream->seek(_offsetFrame2);
+	}
+
+	if (_curFrame == 0)
+		_transparentColourIndex = *(byte *)_surface->getBasePtr(0, 0);
+
+}
+
+const Graphics::Surface *CelDecoder::CelVideoTrack::decodeNextFrame() {
+	// Read chunk
+	/*uint32 frameSize = */ _fileStream->readUint32LE();
+	uint16 frameType = _fileStream->readUint16LE();
+
+	switch (frameType) {
+	case FRAME_TYPE:
+		handleFrame();
+		break;
+	default:
+		error("FlicDecoder::decodeFrame(): unknown main chunk type (type = 0x%02X)", frameType);
+		break;
+	}
+
+	_curFrame++;
+	_nextFrameStartTime += _frameDelay;
+
+	if (_atRingFrame) {
+		// If we decoded the ring frame, seek to the second frame
+		_atRingFrame = false;
+		if (_frameCount == 1) {
+			_fileStream->seek(_offsetFrame1);
+		} else
+			_fileStream->seek(_offsetFrame2);
+	}
+
+	if (_curFrame == 0)
+		_transparentColourIndex = *(byte *)_surface->getBasePtr(0, 0);
+
+	return _surface;
+}
+
+bool CelDecoder::CelVideoTrack::rewind() {
+	// this method is overriden for 2 reasons:
+	// 1) bug in Flic rewind(curFrame)
+	// 2) I changed behaviour of endOfTrack
```

Comment:
This should be fixed now in commit abf17aaa5920572c4bcbeb54577840122e9593fc. So, perhaps this overriden method can be removed now?

Reply:
Unfortunately not, as I've overridden endOfTrack, and in rewind, I need standard endOfTrack.

Reply:
Could you make the comment a bit more descriptive? I was wondering the same thing so it's not self-explanatory right now, especially since (1) is fixed so (2) is the only remaining valid reason if I understand this correctly.